Add Style Guide section for NamedTuple & TypedDict

Solumin opened the pull request:
Resolves #50.

Also adds text to StyleGuide.Classes about using type annotation syntax for attributes.

Solumin committed:
Add Style Guide section for NamedTuple & TypedDict

Resolves #50.

Also adds text to StyleGuide.Classes about using type annotation syntax for attributes.

Solumin committed:
Remove a couple extra blank lines.

rchen152 commented on pep-9999.rst:
```diff
@@ -717,6 +728,30 @@ Maybe::
+    class Point(NamedTuple):
+        x: float
+        y: float
+        def distance(self, other: Point) -> float: ...
```
nit: I'd take out this method, to make the definitions in the "Yes" and "No" sections exactly match, to drive home the point that the class-based syntax should be used even when the same thing can be expressed using the function-based syntax.

srittau commented on pep-9999.rst:
```diff
@@ -559,6 +559,17 @@ No::
     class AnotherError(Exception): pass
 
+Use the type annotation syntax for instance attributes. Class attributes
+should be annotated with ``typing.ClassVar``.
```
"should" -> "must", because I don't think it works otherwise?

srittau commented on pep-9999.rst:
```diff
@@ -717,6 +728,30 @@ Maybe::
+------------------------
+
+Use the class-based syntax for ``typing.NamedTuple`` and
+``typing.TypedDict``, following the Classes section of this style guide.
```
I think this should add an internal hyperlink (no RST expert, though):

```suggestion
``typing.TypedDict``, following the Classes_ section of this style guide.
```

Solumin replied:
I need to look into this more. PEP 12 says, `Every unique section title implicitly defines an internal hyperlink target.` We have 2 sections named "Classes", one in the Style Guide and one in Supported Constructs. I rather like the parallelism of having sections named the same, and I can't think of better names. So I'm not sure if it's possible to link to Style Guide.Classes here. :(

srittau reviewed (approved):
LGTM, two optional comments below.

Solumin committed:
Updates from comments

- should use typing.ClassVar -> must use typing.ClassVar
- removed extra function in NamedTuple class example.

JelleZijlstra commented on pep-9999.rst:
```diff
@@ -559,6 +559,17 @@ No::
     class AnotherError(Exception): pass
 
+Use the type annotation syntax for instance attributes. Class attributes
+must be annotated with ``typing.ClassVar``.
```
I'm actually not sure this is a "must". Most of typeshed doesn't do it (in part because it predates the PEP that added ClassVar), and type checking still mostly works if you get it wrong. I believe the only thing that ClassVar would do in normal classes is to set `instance.universal_thing = "something"`.

srittau replied:
You are right, sorry for the confusion.

Solumin committed:
`must` -> `should` w.r.t. ClassVar

Solumin committed:
Merge branch 'type-stub-pep' into SG-NamedTuple-TypedDict

Solumin committed:
Merge branch 'type-stub-pep' into SG-NamedTuple-TypedDict
